Bugfix/aidumov/dp 72 auth flow dividing

superaydumov opened the pull request:
PR по задаче **DP-72** по разделению флоу авторизации на 3 независимых потока: 
- AuthFlow,
- EditNameFlow,
- JoinSessionAuthFlow.

Создал переиспользуемую вью авторизации ReusableAuthView  для внедрения во вьюконтроллеры.
Добавил методы по взаимодействию с сетью в части создания и изменения пользователя (createUser, updateUser).
Внедрил кастомный лоадер для блокирования UI во время работы с сетью.
Добавил отслеживание активности сетевого соединения через Alamofire.

Close #72 
Close #76

superaydumov committed:
[DP-33] - Добавил RouletteFlow (координатор, презентер, вьюконтроллер, фабрику и необходимые протоколы).
Повесил запуск флоу на нажатие кнопки в CoincidencesViewController (customNavBar).
За запуск сборку RouletteCoordinator отвечает CoincidencesCoordinator.

superaydumov committed:
[DP-33] - Добавил экран онбординга для RouletteFlow, который показывается модально при нажатии на кнопку на экране CoincidencesViewController.
Дополнил файл Resources необходимыми данными.

superaydumov committed:
Merge branch 'develop' into feature/aidumov/DP-33-Roulette-screen

superaydumov committed:
[DP-33] - Перенес метод показа онбордина рулетки в CoincidencesCoordinator.
Обновил логику показа онбординга в CoincidencesPresenter (при наличии 3 совпадений фильмов).
Копка с рулеткой и онбординг появляются если есть 3 и больше совпадений.
Обновил метод finish() в RouletteCoordinator.

superaydumov committed:
[DP-33] - Добавил clearButton в класс CustomButtons.
Заменил кнопку в InvitingUsersViewController

superaydumov committed:
[DP-33] - Внес изменения в класс CustomMovieSelection, чтобы иметь возможность скрывать коллекцию, если в ней нет необходимости.

superaydumov committed:
[DP-33] - Дополнил класс CustomMovieSelection логикой по сокращению текста с информацией о фильме (убрал страны для соответствия макету).
Создал кастомный лэйаут для имитации бесконечной коллекции.
Сверстал коллекцию в классе RouletteViewController (также добавил кастомную ячейку RouletteCollectionViewCell).
Добавил необходимые методы для работы RoulettePresenter.

superaydumov committed:
[DP-33] - Добавил расчет ширины ячейки в зависимости от ширины экрана устройства.

superaydumov committed:
[DP-33] - Создал RouletteStartViewController, добавил управление контроллером в координатор RouletteCoordinator.
Сделал финиш делегатом координатора рулетки координатор сопадений (CoincidencesCoordinator).
Добавил новые вью в RouletteViewController (экран начала рулетки и вью с ожиданием пользователей).
Дополнил методы презентера RoulettePresenter необходимыми методами для управления данными.
Исправил мелкие недочеты в RouletteOnboardingViewController, ReusableUICollectionView (также дополнил методом для выключения отклика на пользователя).
Дополнил файл Resources необходимыми строками.

superaydumov committed:
[DP-33] - Добавил анимацию рулетки.
Добавил методы вьюконтроллера для использования в классе RoulettePresenter.

superaydumov committed:
Merge branch 'develop' into feature/aidumov/DP-33-Roulette-screen

# Conflicts:
#	DavaiPosmotrim.xcodeproj/project.pbxproj
#	DavaiPosmotrim/CoincidencesFlow/CoincidencesCoordinator.swift
#	DavaiPosmotrim/CoincidencesFlow/CoincidencesPresenter.swift
#	DavaiPosmotrim/CoincidencesFlow/CoincidencesViewController.swift

superaydumov committed:
[DP-33] - Перенес переменную totalElements и buffer в вычисляемое свойство moviesCount в классе RoulettePresenter.
Добавил задержку в одну секунду перед появлением онбординга рулетки в классе CoincidencesPresenter.
Отрефакторил методы анимации скролла в классе RouletteViewController.

superaydumov committed:
[DP-33] - Отрефакторил методы анимации скролла в классе RouletteViewController.
Добавил в класс кастомной ячейки RouletteCollectionViewCell метод возвращения id для поиска ячейки в коллекции при скролле.
Дополнил RoulettePresenter публичным свойством для получения id всех фильмов в коллекции.

superaydumov committed:
[DP-33] - Изменил высоту ячейки (зависит от высоты экрана, на котором показывается).

superaydumov committed:
Merge branch 'develop' into feature/aidumov/DP-33-Roulette-screen

superaydumov committed:
[DP-33] - Изменил коллекцию с подключившимися пользователями на соответствующую макету.
Создал кастомную ячейку для переиспользуемой коллекции.
Дополнил RoulettePresenter необходимыми методами и свойствами, связанными с отображением подключившихся пользователей
Добавил в класс RouletteViewController метод для расчета высоты коллекции в зависимости от контента.

superaydumov committed:
[DP-33] - Изменил метод расчета высоты коллекции с именами пользоватей.

superaydumov committed:
[DP-33] - Рефакторинг метода расчета высоты коллекции с пользователями
Сейчас расчет ведется исходя из длины строки в ячейке (имени пользователя).
Обновил метод, иммитирующий подключение пользователей к рулетке.

superaydumov committed:
[DP-33] - Добавил экран с выбранным фильмом в конце рулетки.
Добавил переход на экран с понравившимися фильмами после завершения прогрессбара на кнопке.
Координатор RouletteCoordinator подписан под finishDelegate для перехода в нужную точку флоу.
Обновил класс MatchSelectionMoviesViewController добавив в него инициализируемое свойство для передачи в тайтл кнопки для переиспользования.

superaydumov committed:
[DP-33] - Небольшие исправления констант.

superaydumov committed:
[DP-72] - Добавил все необходимые файлы для EditNameFlow и JoinSessionAuthFlow (координаторы, презентеры, фабрики и протоколы).
Обновил методы создания координаторов в MainCoordinator.

Sbat2000 committed:
[DP-71] добавил сервис для получения жанров

Sbat2000 committed:
[DP-71] поправил настройки безопасности для работы с http

Sbat2000 committed:
[DP-71] настроил инжект сервиса через swinject

Sbat2000 committed:
[DP-71] обновил модель: cover - optional (бывает nil)

Sbat2000 committed:
[DP-71] объеденил сервисы genres and collections

Sbat2000 committed:
[DP-71] добавил метод получения подборок

Sbat2000 committed:
[DP-71] поправил нейминг сервиса (в одном месте был старый)

Sbat2000 committed:
[DP-71] убрал явный return

Sbat2000 committed:
[DP-71] исправил замечания после ревью

superaydumov committed:
[DP-72] - Создал переиспользуемое вью а экраном авторизации/присоединения к сессии/изменения имени для использования внутри нужного флоу.
Добавил переиспользуемое вью в EditNameViewController и JoinSessionAuthViewController.
Убрал соответствие протоколу CoordinatorFinishDelegate в EditNameCoordinator и JoinSessionAuthCoordinator.
Убрал неиспользуемый метод из MainCoordinator.

superaydumov committed:
[DP-72] - добавил свойство в JoinSessionAuhViewController для отслеживания финиша JoinSessionAuthCoordinator.
Убрал ненужные принты.

superaydumov committed:
[DP-72] - Добавил настройку модели переиспользуемой вью в класс EditNameViewController.
Перенес метод calculateCharactersNumber в класс ReusableAuthView. Также отрефакторил методы классе.

superaydumov committed:
[DP-72] - Перенес все необходимые методы в EditNamePresenter и JoinSessionAuthPresenter.
Добавил методы презентеров в метод создания модели переиспользуемого вью setupModel (в классах EditNameViewController и JoinSessionAuthViewController).
Отрефакторил класс ReusableAuthView, убрав ненужные методы и свойства, а также изменив логику.

superaydumov committed:
[DP-72] - Отрефакторил AuthFlow под новую логику (изменил AuthViewController, AuthPresenter и протоколы).
Также отрефакторил метод makeAuthViewController в структуре AuthSceneFactory.

uistelmakh committed:
Merge pull request #74 from DavayPosmotrim/feature/garipov/DP-71-genres-service

Добавил сервисы получения жанров и подборок

superaydumov committed:
[DP-72] - Отрефакторил ReusableAuthView, убрав свойство userName и упростив логику.
Внес необходимые изменения в AuthFlow и EditNameFlow.

superaydumov committed:
[DP-72] - Отрефакторил AuthFlow и EditNameFlow (добавиил нужные методы для работы с сетью, исключил ненужные).
Внес изменения в MainFlow (добавил метод получения пользователя, отрефакторил MainViewController).
Убрал ненужные строки из Resources.

superaydumov committed:
[DP-72] - Добавил проверку количества символов и цифр в коде авторизации к сессии в классе JoinSessionAuthPresenter (отрефакторил метод checkSessionCode).
Добавил отдельный метод downloadSessionCode для загрузки кода авторизации (моковый).
Сейчас авторизация проходит только при условии, что введенный код содержит 4 буквы и 3 цифры, а также совпадает с загруженным из сети кодом.

superaydumov committed:
[DP-33] - Исправлены замечания после ревью.

superaydumov committed:
[DP-33] - Изменил логику работы рулетки в RoulettePresenter (Сейчас отображается 20 рандомных элементов из массива с совпадениями + необходимый фильм. Если в массиве с совпадениями меньше 20 элементов, то отображаются все элементы массива).
Изменил поведение самой рулетки (выбор фильма происходит быстрее).

superaydumov committed:
[DP-33] - Обновил правило для SwiftLint

superaydumov committed:
Merge branch 'develop' into feature/aidumov/DP-33-Roulette-screen

# Conflicts:
#	DavaiPosmotrim.xcodeproj/project.pbxproj

uistelmakh committed:
Merge pull request #73 from DavayPosmotrim/feature/aidumov/DP-33-Roulette-screen

superaydumov committed:
Merge branch 'develop' into bugfix/aidumov/DP-72-AuthFlow-dividing

# Conflicts:
#	DavaiPosmotrim.xcodeproj/project.pbxproj

superaydumov committed:
[DP-72] - Добавил библиотеку ProgressHUD для использования лоадера где необходимо.

IuriiKarpunov committed:
[DP-75] - Интегрировал с сетью: коллекции и жанры

IuriiKarpunov committed:
[DP-75] - устранил замечания ревью

IuriiKarpunov committed:
Merge branch 'develop' into feature/karpunov/DP-75-NetworkIntegration-CreateSessionFlow

# Conflicts:
#	DavaiPosmotrim/UI/ReusableCollection/ReusableUICollectionView.swift

IuriiKarpunov committed:
[DP-75]-замечания ревью

IuriiKarpunov committed:
[DP-75] - после замечаний ревью

uistelmakh committed:
Merge pull request #79 from DavayPosmotrim/feature/karpunov/DP-75-NetworkIntegration-CreateSessionFlow

[DP-75] - Интегрировал с сетью: коллекции и жанры

superaydumov committed:
Merge branch 'develop' into bugfix/aidumov/DP-72-AuthFlow-dividing

# Conflicts:
#	DavaiPosmotrim.xcodeproj/project.pbxproj

superaydumov committed:
[DP-72] - Обновил методы создания и изменения пользователя в AuthFlow и EditNameFlow (методы внутри AuthPresenter и EditNamePresenter).
Имплементировал методы внутри AuthViewController и EditNameViewController.
Добавил кастомный лоадер для блокирования экрана во время выполнения сетевых запросов.
Добавил отслеживание состояния подключения к сети интернет посредством встроенных методов в Alamofire.

SKemenov commented on DavaiPosmotrim/AuthFlow/EditNameFlow/EditNameViewController.swift:
```diff
@@ -0,0 +1,137 @@
+
+    private var reusableAuthModel: ReusableAuthViewModel?
+    private var loadingVC: CustomLoadingViewController?
+    private var checkProperty: Bool?
```
Имхо нейминг не очень удачный. `isServerReachable` или аналогичный.

Насколько корректно тут использовать опциональный `Bool`? Чем он лучше enum в этой ситуации? Имхо от `Bool` ожидаешь только два состояния.

superaydumov replied:
Заменил нейминг.
При использовании `Alamofire` в качестве наблюдателя за состоянием сетевого подключения, необходимо использовать опциональное свойство, потому что в коде метода есть кейс `unknown` (неизвестный для `Alamofire` случай).
В этом случае мы обнуляем наше свойство, отсюда и `nil`.

SKemenov commented on DavaiPosmotrim/AuthFlow/EditNameFlow/EditNameViewController.swift:
```diff
@@ -0,0 +1,137 @@
+    private var reusableAuthModel: ReusableAuthViewModel?
+    private var loadingVC: CustomLoadingViewController?
+    private var checkProperty: Bool?
+    private var networkReachabilityManager: NetworkReachabilityManager?
```
Работа с сетью - это отдельная ответственность. Может стоит вынести в отдельный сетевой сервис?

superaydumov replied:
Done

superaydumov replied:
Согласен.
Done.

SKemenov commented on DavaiPosmotrim/AuthFlow/EditNameFlow/EditNameViewController.swift:
```diff
@@ -0,0 +1,137 @@
+    }
+}
+
+extension EditNameViewController : EditNameViewProtocol {
```
лишний пробел перед `:`

superaydumov replied:
Done

SKemenov commented on DavaiPosmotrim/UI/ReusableAuthView.swift:
```diff
@@ -0,0 +1,301 @@
+        let updatedString = currentString.replacingCharacters(in: range, with: string)
+        
+        switch authEvent {
+        case .joinSession:
+            if updatedString.rangeOfCharacter(from: CharacterSet.alphanumerics.inverted) != nil {
+                return false
+            }
```
`if` внутри `switch` - насколько это оправдано и как потом это поддерживать и сопровождать? 
Может стоит сделать разные вью, используя переиспользуемые компоненты?

superaydumov replied:
Какие вью ты имеешь ввиду?
В этом коде же просто идет проверка на вводимые символы, если мы находимся на экране подключения к сессии

SKemenov replied:
я про это `ReusableAuthView`. В нем столько проверок if и switch

SKemenov commented on DavaiPosmotrim/UI/ReusableCollection/ReusableUICollectionView.swift:
```diff
@@ -98,6 +98,16 @@ final class ReusableUICollectionView: UIView {
+    }
+
+    func updateCollectionView() {
+        DispatchQueue.main.async {
```
Почему тут асинк? 
В методе `loadData()` в `CreateSessionViewController` уже стоит асинк в замыкании - этого недостаточно?

SKemenov replied:
В методе `updateCollection()` тоже стоит асинк - насколько это необходимо?

superaydumov replied:
Done.
Убрал асинк в коде переиспользуемой коллекции.

SKemenov commented on DavaiPosmotrim/UI/ReusableCollection/ReusableUICollectionView.swift.orig:
```diff
@@ -0,0 +1,164 @@
+//
+//  ReusableUICollectionView.swift
```
Странное название файла. Это старая копия? Может стоит удалить?

superaydumov replied:
Done

SKemenov commented on DavaiPosmotrim/AuthFlow/AuthViewController.swift:
```diff
@@ -6,254 +6,131 @@
+    private var reusableAuthModel: ReusableAuthViewModel?
+    private var loadingVC: CustomLoadingViewController?
+    private var checkProperty: Bool?
+    private var networkReachabilityManager: NetworkReachabilityManager?
```
В разных вьюшках одинаковый код работы с сетью?

superaydumov replied:
Перенес в отдельный класс для удобства использования в разных контроллерах

SKemenov commented on DavaiPosmotrim/AuthFlow/AuthPresenter.swift:
```diff
@@ -112,52 +39,29 @@ final class AuthPresenter: AuthPresenterProtocol {
-    }
 
-    func createUser(name: String) {
+    func createUser(name: String, completion: @escaping (Bool) -> Void) {
```
метод `createUser()` - судя по названию - имхо его стоит выносить в отдельный класс для работы с хранением данных (DataStorage?)

superaydumov replied:
На данном этапе предложил бы оставить это в презентере, чтобы не создавать отдельную сущность для хранения данных.
Пока презентер выглядит подходящим местом для работы с сохранением.
Потом можем вернуться к этому вопросу и объединить все взаимодействия с сохранениями в отдельный слой (но уже не в рамках minimum viable product)

SKemenov replied:
ok

SKemenov commented on DavaiPosmotrim/CreateSessionFlow/CreateSessionPresenter.swift:
```diff
@@ -85,7 +87,70 @@ final class CreateSessionPresenter: CreateSessionPresenterProtocol {
+    }
+
+    private func postCreatingSession(segmentIndex: Int) {
+        let dateFormatter = DateFormatter()
+        dateFormatter.dateFormat = "yyyy-MM-dd"
+        let currentDate = dateFormatter.string(from: Date())
```
Работу с форматером можно вынести в `Date` extension

superaydumov replied:
Наверное можно))
Это не мой код. Не могу изменять

SKemenov replied:
ok

SKemenov commented on DavaiPosmotrim/JoinSessionFlow/JoinSessionAuthFlow/JoinSessionAuthViewController.swift:
```diff
@@ -0,0 +1,107 @@
+    var presenter: JoinSessionAuthPresenterProtocol
+
+    private var reusableAuthModel: ReusableAuthViewModel?
+    private var wasDismissedManually = true
```
`isDismissedManually`?

superaydumov replied:
Done.

SKemenov commented on DavaiPosmotrim/MainFlow/MainPresenter.swift:
```diff
@@ -13,10 +13,18 @@ final class MainPresenter: MainPresenterProtocol {
 
+    // MARK: - Private properties
+
+    private let userService: UserServiceProtocol = UserService()
```
Если инициализируем тут, то указание типа `UserServiceProtocol` можно опустить. Или сделать ее переменной и переносить инициализацию переменной в инит:
```swift
init(coordinator: MainCoordinator, userService: UserServiceProtocol = UserService()) {
        self.coordinator = coordinator
        self.userService = userService
    }
```

superaydumov replied:
Done.
Поменял также в EditNamePresenter и AuthPresenter.

SKemenov commented on DavaiPosmotrim/MainFlow/MainPresenter.swift:
```diff
@@ -39,3 +47,20 @@ final class MainPresenter: MainPresenterProtocol {
+
+extension MainPresenter {
+
+    func getUser() {
```
метод `getUser()` - это не ответственность презентера главного экрана

superaydumov replied:
Он сейчас добавлен сюда просто для того, чтобы протестить получение пользователя.
Переместим, когда будем вливать PR Саши в дев.
Это тоже не мой код

SKemenov replied:
принято

SKemenov reviewed (commented):
Эльдар, это большая работа, молодец! 👍 

Рекомендую всё же разбивать PR на меньшие по объему и по логике изменений блоки, по 300-500 строк изменений. Тебе будет легче разделять их логически, ребятам будет легче ревьювить.

Мои предложения написал в комментах по коду.

superaydumov committed:
[DP-72] - Внес исправления после ревью.
Создал отдельный класс проверки состояния сети NetworkReachabilityHandler для исключения дублирования кода во вьюконтроллерах.
Дополнил логику нажатия кнопки "Сохранить" в ReusableAuthView для более корректной и последовательной работы кложеров.

superaydumov commented:
Сергей, спасибо за ревью!
По поводу объема ПР уже говорил ребятам: так вышло из-за разделения на 3 потока с созданием новых файлов.
Иначе бы не вышло, потому что ПРки были бы обрывочными.
Понимаю, что это тяжело смотреть.
